### PR TITLE
EZP-27428: Get translated API properties in the language requested

### PR DESCRIPTION
**Pull request by andrerom**
> Issue: https://jira.ez.no/browse/EZP-27428

### Example

Instead of:
```php
$content = $this->contentService->loadContent(
    42,
    $this->configResolver->getParameter('languages')
);

$name = $content->getVersionInfo()->getName(
    $this->configResolver->getParameter('languages')[0]
);
// OR rather to take care of all languages: $this->translationHelper->getTranslatedContentName($content);

$field = $this->translationHelper->getTranslatedField($content, 'body');
$value = $field->value;
```

We'll get:
```php
$content = $this->contentService->loadContent(
    42,
    $this->configResolver->getParameter('languages')
);

$name = $content->getVersionInfo()->getName();
$value = $content->getFieldValue('body')
```

! However without breaking todays logic in TranslationHelper so existing code works as expected, which is why this can target `master` and not be limited to `7.0`.

### Reviewer notes:
- You might want to review each commit separately here..
- On Content and Version the prioritized field and name language is pre calculated in mapper, this is not strictly needed, it can be done like in the trait for generic name handling in https://github.com/ezsystems/ezpublish-kernel/compare/language_match_api_logic?expand=1#diff-bccf24da09e10e3f51c867eb7d751104R62
  - Reason for not pre-calculated there is same as not reusing pre-calculated prioriticedNameLanguageCode from VersionInfo also in Content: Name might exists in one translation where Field does not. Same goes for name vs description, one or the other might not be translated to a given language.
- Followup on this is PR is #1966

### Todo:
- [ ] Object state (group) assuming the approach taken on content type (group) is ok
- [x] Add test coverage _(currently none to make sure the change does not break anything in existing code)_

### CODE REVIEW

File: eZ/Publish/API/Repository/Values/MultiLanguageValueDescriptionBase.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\API\Repository\Values;
+
+/**
+ * This is the base class for all classes providing translated description logic.
+ *
+ * Provides a uniform way for API consuming logic to generate translated description labels for API objects.
+ */
+abstract class MultiLanguageValueDescriptionBase extends MultiLanguageValueNameBase
```
Comment (Nattfarinn) [inline eZ/Publish/API/Repository/Values/MultiLanguageValueDescriptionBase.php]: I don't like it.

1. `MultiLanguageValueDescriptionBase` is not a child of `MultiLanguageValueNameBase` - not sure if business logic assumption that description always comes with name is a reason to implement inheritance.

2. Both should be interfaces - their purpose is to ensure method signatures on descendants. Interfaces would also resolve multi-inheritance chaining issue.

```php
abstract class ContentType extends ValueObject implements ValueDescriptionAware, ValueNameAware {}
```

vs. 

```php
abstract class MultiLanguageValueNameBase extends ValueObject {}
abstract class MultiLanguageValueDescriptionBase extends MultiLanguageValueNameBase {}
abstract class ContentType extends MultiLanguageValueDescriptionBase {}
```

3. (ad. 1 + 2) If we need a place where we plan to add basic implementation in a future, maybe implementing base, empty abstract aggregating mentioned interfaces is a way to go.

```php
abstract class MultiLanguageValueBase extends ValueObject implements ValueDescriptionAware, ValueNameAware {}
abstract class ContentType extends MultiLanguageValueBase {}
```

File: eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php [hunk 2]
```diff
@@ -25,59 +25,33 @@ public function __construct(array $data = array())
...
-     * 5.x only
-     * This method returns the name of the content type in the given language.
-     *
-     * @param string $languageCode
-     *
-     * @return string the name for the given language or null if none exists.
+     * {@inheritdoc}.
      */
-    public function getName($languageCode)
+    public function getName($languageCode = null)
```
Comment (Nattfarinn) [inline eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php]: Method body will raise Notice in a default case.

File: eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php [hunk 3]
```diff
@@ -25,59 +25,33 @@ public function __construct(array $data = array())
...
-     * 5.x only
-     * This method returns the name of the content type in the given language.
-     *
-     * @param string $languageCode
-     *
-     * @return string the description for the given language or null if none exists.
+     * {@inheritdoc}.
      */
-    public function getDescription($languageCode)
+    public function getDescription($languageCode = null)
```
Comment (Nattfarinn) [inline eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php]: Method body will raise Notice in a default case.

File: eZ/Publish/Core/REST/Client/Values/ContentType/ContentType.php [hunk 4]
```diff
@@ -91,56 +91,33 @@ public function __construct(ContentTypeService $contentTypeService, array $data
...
     /**
-     * This method returns the name of the content type in the given language.
-     *
-     * @param string $languageCode
-     *
-     * @return string the description for the given language or null if none exists.
+     * {@inheritdoc}.
      */
-    public function getDescription($languageCode)
+    public function getDescription($languageCode = null)
```
Comment (Nattfarinn) [inline eZ/Publish/Core/REST/Client/Values/ContentType/ContentType.php]: Method body will raise Notice in a default case.

File: eZ/Publish/Core/REST/Client/Values/ContentType/ContentType.php [hunk 5]
```diff
@@ -91,56 +91,33 @@ public function __construct(ContentTypeService $contentTypeService, array $data
...
     /**
-     * This method returns the name of the content type in the given language.
-     *
-     * @param string $languageCode
-     *
-     * @return string the name for the given language or null if none exists.
+     * {@inheritdoc}.
      */
-    public function getName($languageCode)
+    public function getName($languageCode = null)
```
Comment (Nattfarinn) [inline eZ/Publish/Core/REST/Client/Values/ContentType/ContentType.php]: Method body will raise Notice in a default case.

File: eZ/Publish/API/Repository/Values/MultiLanguageValueNameBase.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ */
+namespace eZ\Publish\API\Repository\Values;
+
+/**
+ * This is the base class for all classes providing translated name logic.
+ *
+ * Provides a uniform way for API consuming logic to generate translated names / labels for API objects, language logic
+ * is meant to also be used for description, fields, .. lookup as well.
+ */
+abstract class MultiLanguageValueNameBase extends ValueObject
```
Comment (Nattfarinn) [inline eZ/Publish/API/Repository/Values/MultiLanguageValueNameBase.php]: (Same as "I don't like it." one)

File: eZ/Publish/Core/Repository/Values/MultiLanguageValueTrait.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\Core\Repository\Values;
+
+/**
+ * @internal Meant for internal use by Repository, type hint against API object instead.
+ */
+trait MultiLanguageValueTrait
```
Comment (Nattfarinn) [inline eZ/Publish/Core/Repository/Values/MultiLanguageValueTrait.php]: Even if internal, wouldn't it be better to split it to two traits? (`MultiLanguageValueNameTrait`, `MultiLanguageValueDescriptionTrait`). Right now it seems to be rather more like a abstract base class than single responsibility trait.

File: eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php (hunk 2 above)
Comment (andrerom) [inline eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php]: This is REST client, it is broken anyway.

Comment (andrerom) [inline eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php]: But yes ,we should add todo's here to make it more clear

File: eZ/Publish/Core/Repository/Values/MultiLanguageValueTrait.php (hunk 7 above)
Comment (andrerom) [inline eZ/Publish/Core/Repository/Values/MultiLanguageValueTrait.php]: could do that, but you'd then duplicate a few of the params.

File: eZ/Publish/API/Repository/Values/ContentType/ContentType.php [hunk 8]
```diff
@@ -179,13 +138,6 @@
     protected $isContainer;
 
     /**
-     * Main language.
-     *
-     * @var string
-     */
-    protected $mainLanguageCode;
```
Comment (andrerom) [inline eZ/Publish/API/Repository/Values/ContentType/ContentType.php]: @alongosz why was this removed? :)

File: eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php (hunk 2 above)
Comment (andrerom) [inline eZ/Publish/Core/REST/Client/Values/ContentType/ContentTypeGroup.php]: Solved now

File: eZ/Publish/API/Repository/Values/ContentType/ContentType.php (hunk 8 above)
Comment (alongosz) [inline eZ/Publish/API/Repository/Values/ContentType/ContentType.php]: @andrerom  classes extending API ContentType redefine it via trait. 
In PHP 5.6 redefining properties [triggers `E_STRICT` error](https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/237973738#L669).

Comment (andrerom) [inline eZ/Publish/API/Repository/Values/ContentType/ContentType.php]: ok, yeah I see it is not there on the other domain types which now uses these traits, so +1 then

File: eZ/Publish/Core/Repository/ContentTypeService.php [hunk 9]
```diff
@@ -168,35 +168,23 @@ public function createContentTypeGroup(ContentTypeGroupCreateStruct  $contentTyp
...
 
     /**
-     * Get a Content Type Group object by id.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException If group can not be found
-     *
-     * @param mixed $contentTypeGroupId
-     *
-     * @return \eZ\Publish\API\Repository\Values\ContentType\ContentTypeGroup
+     * {@inheritdoc}.
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/ContentTypeService.php]: A small nitpick on what I've just discovered. This dot after `{@inheritdoc}` is not needed (and php-cs-fixer is ok with that). When present it becomes a part of doc in the new line. Looks bad in PhpStorm. 
*Applies to several places in the PR.*
inheritdoc_dot

Comment (andrerom) [inline eZ/Publish/Core/Repository/ContentTypeService.php]: Ok, earlier php-cs-fixer was adding it. could you remove it?

**Comment by andrerom** [comment]
Ok, ready for review @alongosz and @Nattfarinn
